### ISSUE TITLE
RNodeInterface: accept a custom byte-channel constructor for phone-attached radios

### ISSUE THREAD

**Issue by emooreatx** [CIRISAI/leviculum#19]
## Goal

Let a host application (mobile or desktop) hand `RNodeInterface` a transparent duplex byte channel instead of a `tokio_serial::SerialStream` opened from a `port_path` string. The radio firmware on the far end still speaks RNode KISS protocol; leviculum still drives detection/configuration/lifecycle; the channel between them is dumb byte-relay.

This unblocks phone-attached LoRa radios on Android (USB host API, BLE GATT) and iOS (BLE GATT) — operating-systems where the wheel never sees `/dev/ttyACM*` and can't `open()` a serial path.

## Current state (v0.7.0 / `reticulum-std/src/interfaces/rnode.rs`)

```rust
pub(crate) struct RNodeInterfaceConfig {
    pub port_path: String,
    // ...
}

pub(crate) fn spawn_rnode_interface(config: RNodeInterfaceConfig) -> InterfaceHandle { ... }

async fn configure_rnode(
    port_path: &str,
    radio: &RadioParams,
) -> Result<(tokio_serial::SerialStream, RNodeDetectResult), RNodeError> { ... }
```

`configure_rnode` opens `tokio_serial::SerialStream::open(&builder)` and returns the stream concretely; the reconnect loop, `read_frames_until_deadline`, `send_radio_config`, and the I/O task all take `&mut tokio_serial::SerialStream`. 15 references total.

## Proposed surface

Add a parallel constructor + public spawn function that takes a **channel factory closure** instead of a path:

```rust
/// A factory the reconnect loop calls to obtain a fresh duplex byte
/// channel. Returns Boxed AsyncRead + AsyncWrite halves so the host
/// app can supply the radio I/O over any substrate (USB-CDC, BLE GATT
/// notification + write characteristic, BT-Classic SPP, mock pipe).
pub trait RNodeChannelFactory: Send + Sync + 'static {
    fn open(&self) -> futures::future::BoxFuture<
        'static,
        Result<(
            Box<dyn tokio::io::AsyncRead  + Send + Unpin>,
            Box<dyn tokio::io::AsyncWrite + Send + Unpin>,
        ), Box<dyn std::error::Error + Send + Sync>>
    >;
}

pub struct RNodeChannelInterfaceConfig {
    pub id: InterfaceId,
    pub name: String,
    pub channel_factory: Arc<dyn RNodeChannelFactory>,
    pub frequency: u32,
    pub bandwidth: u32,
    pub tx_power: u8,
    pub sf: u8,
    pub cr: u8,
    pub st_alock: Option<u16>,
    pub lt_alock: Option<u16>,
    pub flow_control: bool,
    pub buffer_size: usize,
    pub reconnect_notify: Option<mpsc::Sender<InterfaceId>>,
}

pub fn spawn_rnode_channel_interface(
    config: RNodeChannelInterfaceConfig,
) -> InterfaceHandle;
```

## Implementation sketch

1. **Generalize** `configure_rnode`, `read_frames_until_deadline`, `send_radio_config`, `send_*`, `read_io_loop`, etc. from concrete `&mut tokio_serial::SerialStream` to a generic `&mut S where S: AsyncRead + AsyncWrite + Unpin + Send`. tokio's `AsyncReadExt::read` / `AsyncWriteExt::write_all` already work on any such `S`. Type-erase to `Box<dyn ...>` at the spawn boundary so downstream code doesn't have to be generic everywhere.

2. **Refactor** the reconnect loop in `rnode_reconnect_task` to accept EITHER a `port_path: String` (existing behavior — internally calls `tokio_serial::SerialStream::open`) OR a `channel_factory: Arc<dyn RNodeChannelFactory>` (new behavior — internally calls `factory.open().await`). One enum + match handles both.

3. **Keep the existing public surface byte-equal** — `RNodeInterfaceConfig { port_path, ... }` + `spawn_rnode_interface` stays as-is for backwards compat. The new path is purely additive.

4. **Tests**: a `mock_channel_factory` using `tokio::io::duplex(64 * 1024)` and a stub RNode-protocol responder; assert the lifecycle (detect → configure → online → outgoing frame → incoming frame → reconnect-on-drop) matches the existing `test_rnode_interface_lifecycle` ignored-test contract.

## Acceptance

- [ ] `spawn_rnode_channel_interface` lands as public API.
- [ ] Existing `spawn_rnode_interface` keeps its byte-shape contract.
- [ ] A mock-channel test exercises detect + configure + I/O + reconnect against an in-process responder.
- [ ] CHANGELOG entry calls out the additive `*_channel_*` surface + the host-app use case (mobile USB/BLE).

## Consumers

- **CIRISEdge** v7.1.0 will expose `PyEdge.add_rnode_channel_interface(read_cb, write_cb, params)` over this surface, wrapping Python callbacks as an `RNodeChannelFactory` impl.
- **CIRISAgent** will provide the host-app shims: Android USB host, Android BLE, iOS BLE. iOS USB-CDC requires MFi and is deferred.

## Tag

Additive minor — propose `v0.8.0`.

**Comment by emooreatx** [comment]
Shipped in **v0.8.0+ciris.1** (PR #20, commit 8a1a27f). Analysis + implementation by Claude on Eric's behalf.

## What landed

The carrier-agnostic refactor + the public surface:

- `RNodeChannelFactory` trait — `open() -> RNodeChannelOpenFuture` yielding boxed `(AsyncRead, AsyncWrite)` halves (`RNodeChannelHalves`).
- `ReticulumNodeBuilder::add_rnode_channel_interface(factory, frequency, bandwidth, sf, cr, tx_power)`.

Internally: `configure_stream` + the KISS helpers (`detect_on_port`, `read_frames_until_deadline`, `send_radio_config`, `validate_radio_config`) are now generic over `S: AsyncRead + AsyncWrite + Unpin`; the serial open is split into `open_serial_port`; the reconnect loop takes a connector closure (serial opens a `SerialStream`, the channel path `tokio::io::join`s the factory halves).

## Two design choices vs the sketch

1. **Public entry is the builder method, not a bare `pub spawn_rnode_channel_interface`.** Making the spawn fn `pub` would drag the internal `InterfaceHandle` (and `InterfaceInfo`/`InterfaceCounters`/…) into the public API. Every interface in this crate is added via a builder method (`add_rnode_interface`, `add_tcp_client`, …) and the `spawn_*` fns are `pub(crate)`; I kept that idiom. `add_rnode_channel_interface` is exactly what your `PyEdge.add_rnode_channel_interface` wraps. If you'd genuinely rather have the raw spawn fn public, say so and I'll expose it + the handle types.
2. **Factory returns separate read/write halves** (not one `AsyncRead + AsyncWrite` object) because a trait object can name only one non-marker trait. The interface re-joins them with `tokio::io::join`. This also fits BLE naturally (notify characteristic = read, write characteristic = write).

## Pin + usage

```toml
reticulum-std = { git = "https://github.com/CIRISAI/leviculum", tag = "v0.8.0+ciris.1" }
```

```rust
use reticulum_std::interfaces::{RNodeChannelFactory, RNodeChannelHalves, RNodeChannelOpenFuture};

struct PhoneRadio { /* USB/BLE handles, or Python read/write callbacks */ }
impl RNodeChannelFactory for PhoneRadio {
    fn open(&self) -> RNodeChannelOpenFuture {
        Box::pin(async move {
            let (r, w) = /* open a fresh duplex channel to the radio */;
            Ok((Box::new(r) as _, Box::new(w) as _))
        })
    }
}

let node = ReticulumNodeBuilder::new()
    .add_rnode_channel_interface(Arc::new(PhoneRadio{..}), 868_000_000, 125_000, 7, 5, 17)
    .build().await?;
```

Test proving it end-to-end: `test_rnode_channel_interface_lifecycle` — detect → configure → online → outbound → inbound over `tokio::io::duplex`, no serial port. Adoption tracked in CIRISEdge#195.

**Comment by emooreatx** [comment]
Thanks for the merge + tag — `+ciris.1` build metadata is the right call for fork-marking, and the `RNodeChannelFactory` trait shape matches what edge needs.

One follow-up: **edge needs the raw `spawn_rnode_channel_interface` exposed `pub`** alongside (or instead of) `ReticulumNodeBuilder::add_rnode_channel_interface`.

## Why

The whole point of edge's `PyEdge.add_rnode_channel_interface(name, read_cb, write_cb, params) -> handle` ([CIRISEdge#219](https://github.com/CIRISAI/CIRISEdge/issues/219)) is **runtime attach** — Python calls it any time after `init_edge_runtime` to plug a radio in. The builder pattern implies construction-time wiring; with builder-only:

1. Agent must open the USB/BLE radio FIRST.
2. Then call `init_edge_runtime` with the channel factory passed in as a startup arg.
3. Cannot detach the radio without tearing down the whole `ReticulumNode` and rebuilding (kills every other interface in the process — TCP, HTTPS bootstrap, etc.).

The hot-plug semantics matter for the Android USB-OTG / iOS BLE shims — the user might plug the radio in mid-session, or swap radios, or have the radio attached only when in motion.

## What edge needs

```rust
/// Spawn an RNode interface backed by a host-supplied byte-channel
/// factory + register it on an already-running ReticulumNode.
/// Returns the InterfaceHandle so the caller can manage the interface's
/// lifecycle (drop = detach).
pub fn spawn_rnode_channel_interface(
    node: &ReticulumNode,                    // or `&self` if it lives on ReticulumNode
    config: RNodeChannelInterfaceConfig,
) -> InterfaceHandle;
```

The "leaking InterfaceHandle" concern is actually load-bearing for edge: PyEdge holds the handle on the returned Python object, and `__del__` / `.close()` drops it to tear the interface down cleanly. Same lifecycle pattern as `PyDurableHandle` / `PySubscriptionHandle` in v7.x.

## Surface preserves builder

The builder method stays useful for callers that DO want construction-time wiring (CLI tools, daemons with a fixed config). The raw spawn fn is the runtime-attach companion. Keeping both = no regression for current users + edge unblocked.

## Tag

Either a `v0.8.1+ciris.1` (additive minor) or rolled into the next cut. Edge's [CIRISEdge#195 / #219](https://github.com/CIRISAI/CIRISEdge/issues/219) adoption tracker gates on this.

Happy to validate a pre-release tag against the PyEdge shim.

**Comment by emooreatx** [comment]
Runtime hot-plug shipped in **v0.8.1+ciris.1** (PR #21, commit eb32cf1) — this is the raw runtime-attach you need, done as a drop-to-detach lifecycle handle.

```toml
reticulum-std = { git = "https://github.com/CIRISAI/leviculum", tag = "v0.8.1+ciris.1" }
```

## API

```rust
let handle: RNodeChannelHandle = node.spawn_rnode_channel_interface(RNodeChannelConfig {
    factory: Arc::new(my_radio),   // Arc<dyn RNodeChannelFactory>
    frequency: 868_000_000, bandwidth: 125_000, sf: 7, cr: 5, tx_power: 17,
    st_alock: None, lt_alock: None, flow_control: false,
    buffer_size: /* RNODE_DEFAULT_BUFFER_SIZE */ 64,
})?;
// hold `handle` => radio attached; drop it (or handle.detach()) => detached.
```

- Attach works on a **running** node, from any thread (the interface task spawns on the node's own runtime — fine to call after `init_edge_runtime` from a Python thread). Returns `Error::NotRunning` if called before `start()`.
- `RNodeChannelHandle` is the lifecycle token: **hold = attached, drop/`.detach()` = detached**, torn down cleanly (same path a disconnecting TCP interface takes). PyEdge stores it on the Python object and drops it in `close()`/`__del__`.
- `node.spawn_rnode_channel_interface(...)` ≈ your proposed free fn `spawn_rnode_channel_interface(node, config)` (it's `&self`).

## The one deviation from your spec, and why
Return type is `RNodeChannelHandle`, **not** the internal `InterfaceHandle`. I verified in the driver that the raw `InterfaceHandle`'s I/O channels must be **owned by the event loop** for the interface to route at all — you can't both run the interface and hold its handle. `RNodeChannelHandle` gives the exact semantics you described (hold/drop lifecycle) while keeping the interface functional. If you specifically need a field off the raw handle (counters, etc.), say which and I'll expose an accessor — but the lifecycle (your stated need) is fully covered.

The construction-time `ReticulumNodeBuilder::add_rnode_channel_interface` from v0.8.0+ciris.1 remains for startup wiring; this is the runtime counterpart.

Test proving it: `test_runtime_attach_detach_rnode_channel` — starts a node, attaches over an in-memory channel, runs detect→configure (attach), drops the handle, asserts the channel closes (detach). Adoption: CIRISEdge#219 / v7.2.0.
